NavigationRail, NavigationBar, NavigationDrawer controls for MaterialDesign3

StefanoRivolta-Previero opened the pull request:
# Summary

From issue #2578 

I've added the Navigation Rail, Navigation Bar and Navigation Drawer controls as styles for the ListBox. This is probably not ideal, as it had already been pointed out, as it limits some of the things that can be done, but it was also somewhat easier for me to do rather than creating a completely new CustomControl. In the future that may be a better way to implement the controls though.

The messy part has to do with the Icons, since it needs both a selected version and an unselected (outlined) version. The best way I came up with was to make them a part of the Style, so it works like this:

```
<ListBox.ItemContainerStyle>
    <Style TargetType="ListBoxItem" BasedOn="{StaticResource MaterialDesign3NavigationDrawerPrimaryListBoxItem}">
        <Setter Property="materialDesign:NavigationDrawerAssist.SelectedIcon" Value="{Binding SelectedIcon}"/>
        <Setter Property="materialDesign:NavigationDrawerAssist.UnselectedIcon" Value="{Binding UnselectedIcon}"/>
    </Style>
</ListBox.ItemContainerStyle>
```
Or this without MVVM:

```
<ListBoxItem>
    <ListBoxItem.Style>
        <Style TargetType="ListBoxItem" BasedOn="{StaticResource MaterialDesign3NavigationRailPrimaryListBoxItem}">
            <Setter Property="materialDesign:NavigationRailAssist.SelectedIcon" Value="Circle"/>
            <Setter Property="materialDesign:NavigationRailAssist.UnselectedIcon" Value="CircleOutline"/>
            <Setter Property="materialDesign:NavigationRailAssist.IconSize" Value="16"/>
        </Style>
    </ListBoxItem.Style>
    <TextBlock Text="text 1"/>
</ListBoxItem>
```

I still have to add the Badge to the items, as per spec. Waiting for feedback before I do. I also haven't re-implemented the TabControl style for NavigationRail, should I?

# Navigation Rail

Specs: https://m3.material.io/components/navigation-rail/overview

![sQvUiT4VkE](https://user-images.githubusercontent.com/58171461/160871598-ca6c5a7a-3659-4923-8154-949301ac05b3.png)

# Navigation Bar

Specs: https://m3.material.io/components/navigation-bar/overview

![EASLg7jgj8](https://user-images.githubusercontent.com/58171461/160871831-8b1a0022-5947-4316-8f82-f7a299eef28f.png)

# Navigation Drawer

Specs: https://m3.material.io/components/navigation-drawer/overview

![7DRbW1TXJH](https://user-images.githubusercontent.com/58171461/160872085-6684b4fc-d62c-42de-b93f-7fc7492e816f.png)

# Demo App

I've updated the DemoApp to use NavigationRail, NavigationBar and NavigationDrawer in relation to the size of the window, similarly to this: https://m3.material.io/foundations/adaptive-design/overview

![devenv_WyfVKEp0fH](https://user-images.githubusercontent.com/58171461/160873953-cf7276f3-28ef-4baf-867f-0ab9b027121f.gif)

I'll wait for feedback. 
By the way, should we differentiate the *Assist.cs files between MD2 and MD3 or should we leave them the same?

Keboo commented on MaterialDesignThemes.Wpf/MaterialDesignThemes.Wpf.csproj:
```diff
@@ -27,4 +27,17 @@
   <ItemGroup>
     <ProjectReference Include="..\MaterialDesignColors.Wpf\MaterialDesignColors.Wpf.csproj" />
   </ItemGroup>
+  <ItemGroup>
+    <Page Update="Themes\MaterialDesign3.NavigationBar.xaml">
+      <XamlRuntime>$(DefaultXamlRuntime)</XamlRuntime>
+      <SubType>Designer</SubType>
+    </Page>
+    <Page Update="Themes\MaterialDesign3.NavigationDrawer.xaml">
+      <XamlRuntime>$(DefaultXamlRuntime)</XamlRuntime>
+      <SubType>Designer</SubType>
+    </Page>
+    <Page Update="Themes\MaterialDesign3.NavigationRail.xaml">
+      <XamlRuntime>$(DefaultXamlRuntime)</XamlRuntime>
+    </Page>
+  </ItemGroup>
```
I believe these lines can be omitted.

Keboo commented on MaterialDesign3.Demo.Wpf/Domain/SampleItem.cs:
```diff
@@ -0,0 +1,10 @@
+
+public class SampleItem
+{
+    public string Title { get; set; }
```
Because nullable reference types is enabled, this should be declared as a nullable string (this is the build failure on the pipeline; only a warning when compiled locally)
```suggestion
    public string? Title { get; set; }
```

Keboo reviewed (commented):
First of all, let me start by saying thank you. This work looks amazing!

I left a few small comments, but I think this is very close to being merged. 
As for your question on the *Assist classes. My preference would be to keep them together. Attached properties are difficult to discover as it is, so I would be hesitant to separate them out. With that said, I suspect there may be cases where an attached property only applies to MD2 or MD3. For those I think we will just need to evaluate on a case-by-case basis.

StefanoRivolta-Previero commented:
I've fixed the errors and added the badge to the navigation controls. 
To pass through the properties of the badge, I had to create a badge assist class, not sure if it's the right way to do it though.
I also added a new style for the badged control, so that it only shows the small dot, no text.

Some changes also to the Demo App. To showcase the badge I've added a Notifications property to the DemoItem class and a little control to increase or dismiss the notifications. Also new an extended FAB and a collapse button in the drawer when in standard mode.

![devenv_JxZR7mTPYK](https://user-images.githubusercontent.com/58171461/161556621-f98620ed-79e4-41de-8b41-e9ac28cc3494.gif)

BigBadBleuCheese commented:
Where is this groovy new demo? I downloaded the 4.5.0 demo and it seems mostly as it was.

Keboo commented:
@BigBadBleuCheese it lives under the [MaterialDesign3.Demo.Wpf directory in the repo](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/tree/master/MaterialDesign3.Demo.Wpf). However your comments reminds me that it has not been added to the pipeline which is why it is not currently being included in the release artifacts. I will work on getting that added, because it is pretty cool.
